src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -360,9 +360,36 @@ private static string[] HandleClear(string[] input)
             return input;
         }
 
-        // Trying to fetch a property value from tfm property bags.
-        // If defined the property should have identical values in all of the occurances.
-        private static TValue GetNonEvaluatedPropertyOrNull<TValue>(
+        private static TValue GetSingleOrDefaultProperties<TValue>(
+            IVsTargetFrameworks tfms,
+            string propertyName,
+            Func<string, TValue> valueFactory)
+        {
+            var warningProperties = GetNonEvaluatedPropertyOrNull(tfms, propertyName, valueFactory);
+
+            if (warningProperties.Count() > 1)
+            {
+                return default(TValue);
+            }
+            else
+            {
+                return warningProperties.SingleOrDefault();
+            }
+
+        }
+
+        private static IEnumerable<NuGetLogCode> GetSingleOrDefaultNuGetLogCodes(
+            IVsTargetFrameworks tfms,
+            string propertyName,
+            Func<string, IEnumerable<NuGetLogCode>> valueFactory)
+        {
+            var logCodeProperties = GetNonEvaluatedPropertyOrNull(tfms, propertyName, valueFactory);
+
+            return MSBuildStringUtility.GetSingleOrDefaultDistinctNuGetLogCodes(logCodeProperties);
+        }
+
+        // Trying to fetch a list of property value from all tfm property bags.
+        private static IEnumerable<TValue> GetNonEvaluatedPropertyOrNull<TValue>(
```

Comment:
nit: GetNonEvaluatedPropertyValues. Will it ever return null?

Reply:
it will return enumerable which contains null